distributed/scheduler.py

```diff
@@ -1590,6 +1596,36 @@ def get_nbytes(self, stream=None, keys=None, summary=True):
 
             return result
 
+    def publish_dataset(self, stream=None, keys=None, data=None, name=None,
+                        client=None):
+        for k in keys:
+            self.who_wants[k].add('published-%s' % name)
+        self.published_data[name] = {'data': data, 'keys': keys,
+                                     'clients': {client}}
+
+    def unpublish_dataset(self, stream=None, name=None):
+        out = self.published_data.pop(name, {'keys': []})
+        for k in out['keys']:
+            self.who_wants[k].remove('published-%s' % name)
```

Comment:
> On who_wants/wants_what, this to ensure consistency for future-proofing, rather than any material difference at this moment? 

Various parts of the code assume that those two dicts are always in sync.  They should be considered a single data structure that is somewhat cumbersome to update.  I would expect some things to err if these get out of sync.  The fact that things aren't erring is somewhat concerning.  I should add checks here into the runtime validation methods.